extension/src/functions.rs

```diff
@@ -0,0 +1,136 @@
+use super::*;
+use core::fmt::Debug;
+pub use decoding::{Decode, Decodes, Processor};
+pub use matching::{Equals, FunctionId, Matches};
+use pallet_contracts::chain_extension::{BufIn, BufOut};
+
+/// A chain extension function.
+pub trait Function {
+	/// The configuration of the contracts module.
+	type Config: pallet_contracts::Config;
+
+	/// Executes the function.
+	///
+	/// # Parameters
+	/// - `env` - The current execution environment.
+	fn execute<E: Ext<T = Self::Config>, S: BufIn + BufOut>(
+		env: &mut Environment<E, S>,
+	) -> Result<RetVal>;
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(1, 3)]
+#[tuple_types_custom_trait_bound(Function + Matches)]
+impl<Runtime: pallet_contracts::Config> Function for Tuple {
+	for_tuples!( where #( Tuple: Function<Config=Runtime> )* );
+	type Config = Runtime;
+	fn execute<E: Ext<T = Self::Config>, S: BufIn + BufOut>(
+		env: &mut Environment<E, S>,
+	) -> Result<RetVal> {
+		// Attempts to match a specified extension/function identifier to its corresponding function, as configured by the runtime.
+		for_tuples!( #(
+            if Tuple::matches(&env) {
+                return Tuple::execute(env)
+            }
+        )* );
+
+		// Otherwise returns error indicating an unmatched request.
+		Err(pallet_contracts::Error::<Self::Config>::NoChainExtension.into())
```

Comment:
Same error raised when no active chain extension is found: https://github.com/paritytech/polkadot-sdk/blob/f239abac93b8354fec1441e39727b0706b489415/substrate/frame/contracts/src/wasm/runtime.rs#L2258

Might be okay to start, probably should be configurable.

Reply:
This prevents adding the `UnknownCall`. In the end it is a u32 on the contract side which is turned into a clear error. Docs can support what this error means. It prevents the situation where we end up wanting to remove the previously used `UnknownCall` so sounds like a good plan.

Reply:
Regarding this point, there is also [DecodingFailed](https://github.com/paritytech/polkadot-sdk/blob/717bbb24c40717e70d2e3b648bcd372559c71bd2/substrate/frame/contracts/src/lib.rs#L1244).

Reply:
Changed to `DecodingFailed` in latest commit as the error description is roughly closer to the reason for the error. Note also that we have DecodingFailed in the contract-side error, with an explicit mapping to this variant when it occurs in the runtime keeping everything aligned.

In reality, as long as devs are using the pop-api crate, they should never experience this error - only those who are implementing the API.